### ISSUE TITLE
T-SQL ELSE/ELSE IF indentation issue: Keywords misclassified as Word tokens causing LT02 violations

### ISSUE THREAD

**Issue by Fank** [quarylabs/sqruff#1873]
# T-SQL ELSE/ELSE IF Indentation Issue: Keywords Misclassified as Word Tokens

## Summary

The LT02 (indent) rule incorrectly flags T-SQL `ELSE` and `ELSE IF` keywords as requiring indentation when they should be at the same level as the `IF` keyword. This affects the `test_pass_tsql_else_if` test case and breaks proper T-SQL control flow formatting.

## Root Cause (Confirmed via Debug Tracing)

**Critical Finding**: T-SQL's word-aware parsing system incorrectly classifies `ELSE` keywords as `SyntaxKind::Word` instead of `SyntaxKind::Keyword`, preventing the LT02 rule from recognizing them as part of the control flow structure.

### Debug Evidence

**T-SQL (Failing) - Block classification:**
```
Block 0: [Keyword] | depth: 4      (IF keyword - CORRECT)
Block 18: [Word] | depth: 5        (ELSE keyword - WRONG!)
```

**BigQuery (Working) - Block classification:**
```
Block 0: [Keyword] | depth: 3      (IF keyword - CORRECT)
Block 12: [Keyword] | depth: 3     (ELSE keyword - CORRECT!)
```

The LT02 rule's indentation logic expects control flow keywords to be classified as `SyntaxKind::Keyword` to properly recognize indentation structure. When `ELSE` appears as `SyntaxKind::Word`, the rule fails to identify it as part of the `IF` statement's control structure.

### Expected (Should Pass)
```sql
IF (1 > 1)
    PRINT 'A';
ELSE IF (2 > 2)
    PRINT 'B';
ELSE IF (3 > 3)
    PRINT 'C';
ELSE
    PRINT 'D';
```

### Actual (LT02 Violations)
```
L:   3 | P:   1 | LT02 | Expected indent of 4 spaces [layout.indent]
L:   5 | P:   1 | LT02 | Expected indent of 4 spaces [layout.indent]  
L:   7 | P:   1 | LT02 | Expected indent of 4 spaces [layout.indent]
```

## Test Case Details

- **Failing Test**: `test_pass_tsql_else_if` in `crates/lib/test/fixtures/rules/std_rule_cases/LT02-indent.yml`
- **Configuration**: `dialect = tsql`, `allow_implicit_indents = true`
- **Location**: Lines 3, 5, 7 (all ELSE keywords)

### Confirmed Working Elements
- ✅ **Parser Structure**: Both `IfStatementSegment` and `WordAwareIfStatementSegment` have identical structure to working BigQuery parser
- ✅ **No Parsing Errors**: `--parsing-errors` flag confirms both parsers generate valid AST
- ✅ **Indent/Dedent Placement**: `MetaSegment::indent()` and `MetaSegment::dedent()` correctly placed before ELSE clauses
- ✅ **Configuration**: `allow_implicit_indents` setting has no effect on the issue

### Root Cause Analysis  
The issue is **NOT** in:
- Parser structure (matches BigQuery exactly)
- Missing indent/dedent metadata
- Configuration settings
- Parsing errors

The issue **IS** in:
- Token classification pipeline where `ELSE` keywords become `Word` tokens instead of `Keyword` tokens
- LT02 rule's expectation that control flow keywords are classified as `SyntaxKind::Keyword`

### Affected Components
- **File**: `crates/lib-dialects/src/tsql.rs`
- **Parsers**: `IfStatementSegment` (line ~1791), `WordAwareIfStatementSegment` (line ~10214)
- **Rule**: `crates/lib/src/rules/layout/lt02.rs` → `crates/lib/src/utils/reflow/reindent.rs`

### Architecture Impact
T-SQL's word-aware parsing system, designed to handle complex procedure bodies and dynamic SQL, appears to interfere with proper keyword classification for control flow tokens. This creates a conflict between:
1. **Dialect Complexity**: Word-aware parsing for T-SQL-specific features
2. **Rule Expectations**: LT02 expecting standard keyword classification

### Parser Priority Fix (Partial Success)
Modified `WordAwareIfStatementSegment` to prioritize keyword tokens:
```rust
// Changed from:
StringParser::new("ELSE", SyntaxKind::Word),
Ref::keyword("ELSE")

// To:
Ref::keyword("ELSE"), 
StringParser::new("ELSE", SyntaxKind::Word),
```

**Result**: Parser structure improved but tokens still classified as `[Word]` in debug output.

### Working Dialects
- **BigQuery**: `IF...THEN...ELSEIF...ELSE...END IF` - ✅ No LT02 errors, proper keyword classification
- **ANSI**: Standard control flow works correctly

### T-SQL Specific Issues
- **Two-keyword syntax**: `ELSE IF` (two keywords) vs `ELSEIF` (single keyword)
- **Word-aware parsing**: Complex token classification pipeline
- **Procedure context**: Additional parsing layers for stored procedures

## Required Fix

The fix requires investigation into T-SQL's lexer/parser pipeline to ensure `ELSE` keywords maintain `SyntaxKind::Keyword` classification throughout AST processing, particularly in:

1. **Token Classification**: Ensure lexer properly identifies ELSE as keyword
2. **Parser Precedence**: Verify `IfStatementSegment` vs `WordAwareIfStatementSegment` matching order
3. **AST Processing**: Prevent downstream conversion of keywords to words

### Primary
- `crates/lib-dialects/src/tsql.rs` - Dialect definition and parsers  
- `crates/lib-dialects/src/tsql_keywords.rs` - Keyword definitions
- `crates/lib/src/utils/reflow/reindent.rs` - LT02 indentation logic

### Test Files
- `crates/lib/test/fixtures/rules/std_rule_cases/LT02-indent.yml` - Failing test case
- Test case: `test_pass_tsql_else_if`

## Minimal Reproduction

```sql
IF (1 > 1)
    PRINT 'A';
ELSE
    PRINT 'B';
```

**Command**: `sqruff lint --dialect tsql file.sql`
**Expected**: No LT02 violations  
**Actual**: `L: 3 | P: 1 | LT02 | Expected indent of 4 spaces`

## Impact

- **Severity**: High - Affects basic T-SQL control flow formatting
- **Scope**: All T-SQL IF/ELSE IF/ELSE statements
- **Users**: Anyone using sqruff with T-SQL dialect for control flow statements

**Comment by Fank** [comment]
Hello, I am Gemini, a large language model from Google. I have been assisting with the analysis of this LT02 indentation issue for the T-SQL dialect, and I have synthesized the following insights based on a detailed review of the parser implementation and related research documents.

### Executive Summary

The root cause of this issue is **not a bug in the T-SQL parser**. The parser correctly generates a nested Abstract Syntax Tree (AST) with the appropriate `indent` and `dedent` markers for control flow statements like `IF...ELSE IF...ELSE`.

The problem lies in the **LT02 rule's interpretation of the AST produced by the T-SQL dialect**. Specifically, the rule fails to correctly identify control flow keywords because they are classified as `SyntaxKind::Word` instead of `SyntaxKind::Keyword` in procedural contexts.

### Detailed Analysis

1.  **T-SQL's "Word-Aware" Parsing is Working as Designed**: The T-SQL lexer intentionally tokenizes keywords like `IF`, `ELSE`, `BEGIN`, and `END` as generic `word` tokens within procedural blocks. The parser is designed to handle this by matching the string content of these words, and it successfully builds a structurally correct AST.

2.  **The Real Root Cause**: Debug tracing confirms that when the LT02 rule inspects the AST for a T-SQL `IF` statement, it sees the `ELSE` token as a `[Word]`. The rule's logic, however, expects a `[Keyword]` to recognize it as part of the indentation structure. Because it sees a `Word`, it treats it as a new, separate statement that requires indentation, leading to the incorrect linting error.

3.  **Parser Structure is Sound**: The T-SQL `IfStatementSegment` parser has the correct structure, with `MetaSegment::dedent()` placed *before* the `ELSE` and `ELSE IF` clauses. This structure is identical to the one used in the BigQuery dialect, which does not suffer from this issue because its control flow tokens are consistently classified as `Keyword`s.

4.  **Broader Impact**: This is not an isolated problem with `ELSE IF`. The same root cause affects other T-SQL constructs, including `BEGIN...END` blocks and function definitions, leading to widespread indentation and formatting issues in the T-SQL dialect.

### Recommendation

The most effective and lowest-risk solution is to **modify the LT02 rule**, not the T-SQL parser.

The parser is correctly handling the complexities of the T-SQL language. The fix should be to make the LT02 rule "dialect-aware" enough to handle the AST that the T-SQL parser produces. The rule should be updated to recognize control flow constructs by checking for either:

*   A `SyntaxKind::Keyword` token.
*   A `SyntaxKind::Word` token whose raw text matches a control flow word (e.g., `token.raw_text().eq_ignore_ascii_case("else")`).

This approach addresses the problem at the correct layer—the rule interpretation—and avoids a high-risk, complex rewrite of the core parsing engine.

**Comment by Fank** [comment]
Hello, this is an updated and more detailed analysis from Gemini, a large language model from Google. I am expanding on my previous comment to include a discussion of potential solutions and their respective trade-offs to provide a comprehensive view of the issue.

### Executive Summary

The root cause of this issue is **not a bug in the T-SQL parser**. The parser correctly generates a nested Abstract Syntax Tree (AST) with the appropriate `indent` and `dedent` markers for control flow statements like `IF...ELSE IF...ELSE`.

The problem lies in the **LT02 rule's interpretation of the AST produced by the T-SQL dialect**. Specifically, the rule fails to correctly identify control flow keywords because they are classified as `SyntaxKind::Word` instead of `SyntaxKind::Keyword` in procedural contexts.

### Detailed Analysis

1.  **T-SQL's \"Word-Aware\" Parsing is Working as Designed**: The T-SQL lexer intentionally tokenizes keywords like `IF`, `ELSE`, `BEGIN`, and `END` as generic `word` tokens within procedural blocks. The parser is designed to handle this by matching the string content of these words, and it successfully builds a structurally correct AST.

2.  **The Real Root Cause**: Debug tracing confirms that when the LT02 rule inspects the AST for a T-SQL `IF` statement, it sees the `ELSE` token as a `[Word]`. The rule's logic, however, expects a `[Keyword]` to recognize it as part of the indentation structure. Because it sees a `Word`, it treats it as a new, separate statement that requires indentation, leading to the incorrect linting error.

3.  **Parser Structure is Sound**: The T-SQL `IfStatementSegment` parser has the correct structure, with `MetaSegment::dedent()` placed *before* the `ELSE` and `ELSE IF` clauses. This structure is identical to the one used in the BigQuery dialect, which does not suffer from this issue because its control flow tokens are consistently classified as `Keyword`s.

4.  **Broader Impact**: This is not an isolated problem with `ELSE IF`. The same root cause affects other T-SQL constructs, including `BEGIN...END` blocks and function definitions, leading to widespread indentation and formatting issues in the T-SQL dialect.

### Potential Solutions and Their Trade-offs

There are two high-level approaches to solving this problem: either make the rules smarter or make the parser produce a "perfect" AST.

#### Approach 1: Enhance the Linting Rules (Recommended)

This approach involves modifying the affected rules (like LT02) to be aware of the T-SQL dialect's specific AST structure.

*   **How**: The rule would be updated to recognize control flow constructs by checking for either a `SyntaxKind::Keyword` token OR a `SyntaxKind::Word` token whose raw text matches a control flow word (e.g., `token.raw_text().eq_ignore_ascii_case(\"else\")`).
*   **Pros**:
    *   **Low Risk & Targeted**: Directly solves the user-facing problem (incorrect linting) with minimal changes to the core engine.
    *   **Pragmatic**: Acknowledges and works with the existing, deliberate architectural trade-offs of the T-SQL dialect.
*   **Cons**:
    *   **Rule Complexity**: The linting rules become slightly more complex as they need to handle dialect-specific variations in the AST.

#### Approach 2: Re-architect the Parser to Produce "Perfect" Keywords

This approach aims to fix the issue at the source, ensuring the parser *always* outputs a `SyntaxKind::Keyword` for control flow words. This would involve a major architectural change.

*   **Sub-option 2a: Lexer Feedback Loop**: The parser would need to send signals back to the lexer to change its state mid-parse (e.g., "we are now in a procedure, treat `IF` as a keyword").
    *   **Pros**: Would produce a "pure" AST, simplifying the rules.
    *   **Cons**: **Extremely high risk and complexity**. This creates a "lexer hack" that is difficult to design, debug, and maintain, and could introduce performance overhead. It would require a significant rewrite of the core parsing engine.

*   **Sub-option 2b: AST Post-Processing**: A second pass would walk the generated AST and mutate it, changing `Word` tokens to `Keyword` tokens based on their context.
    *   **Pros**: Lower risk than a feedback loop as it isolates the logic.
    *   **Cons**: Adds a performance penalty due to the second full pass over the AST and introduces redundant logic that is already present in the word-aware parser rules.

### Final Recommendation

The most effective and lowest-risk solution is **Approach 1: Modify the LT02 rule**.

The parser is correctly handling the complexities of the T-SQL language. The fix should be to make the LT02 rule "dialect-aware" enough to handle the AST that the T-SQL parser produces. This approach addresses the problem at the correct layer—the rule interpretation—and avoids a high-risk, complex rewrite of the core parsing engine.

**Comment by Fank** [comment]
Hello, I am Gemini, a large language model from Google. This comment provides a single, consolidated, and detailed analysis of the LT02 indentation issue for the T-SQL dialect, superseding any previous comments.

### Executive Summary

The root cause of this issue is **not a bug in the T-SQL parser**. The parser correctly generates a nested Abstract Syntax Tree (AST) with the appropriate `indent` and `dedent` markers for control flow statements like `IF...ELSE IF...ELSE`.

The problem lies in the **LT02 rule's interpretation of the AST produced by the T-SQL dialect**. Specifically, the rule fails to correctly identify control flow keywords because they are classified as `SyntaxKind::Word` instead of `SyntaxKind::Keyword` in procedural contexts.

### Detailed Analysis

1.  **T-SQL's \"Word-Aware\" Parsing is Working as Designed**: The T-SQL lexer intentionally tokenizes keywords like `IF`, `ELSE`, `BEGIN`, and `END` as generic `word` tokens within procedural blocks. The parser is designed to handle this by matching the string content of these words, and it successfully builds a structurally correct AST.

2.  **The Real Root Cause**: Debug tracing confirms that when the LT02 rule inspects the AST for a T-SQL `IF` statement, it sees the `ELSE` token as a `[Word]`. The rule's logic, however, expects a `[Keyword]` to recognize it as part of the indentation structure. Because it sees a `Word`, it treats it as a new, separate statement that requires indentation, leading to the incorrect linting error.

3.  **Parser Structure is Sound**: The T-SQL `IfStatementSegment` parser has the correct structure, with `MetaSegment::dedent()` placed *before* the `ELSE` and `ELSE IF` clauses. This structure is identical to the one used in the BigQuery dialect, which does not suffer from this issue because its control flow tokens are consistently classified as `Keyword`s.

4.  **Broader Impact**: This is not an isolated problem with `ELSE IF`. The same root cause affects other T-SQL constructs, including `BEGIN...END` blocks and function definitions, leading to widespread indentation and formatting issues in the T-SQL dialect.

### Potential Solutions and Their Trade-offs

There are two high-level approaches to solving this problem: either make the rules smarter or make the parser produce a "perfect" AST.

#### Approach 1: Enhance the Linting Rules (Recommended)

This approach involves modifying the affected rules (like LT02) to be aware of the T-SQL dialect's specific AST structure.

*   **How**: The rule would be updated to recognize control flow constructs by checking for either a `SyntaxKind::Keyword` token OR a `SyntaxKind::Word` token whose raw text matches a control flow word (e.g., `token.raw_text().eq_ignore_ascii_case(\"else\")`).
*   **Pros**:
    *   **Low Risk & Targeted**: Directly solves the user-facing problem (incorrect linting) with minimal changes to the core engine.
    *   **Pragmatic**: Acknowledges and works with the existing, deliberate architectural trade-offs of the T-SQL dialect.
*   **Cons**:
    *   **Rule Complexity**: The linting rules become slightly more complex as they need to handle dialect-specific variations in the AST.

#### Approach 2: Re-architect the Parser to Produce \"Perfect\" Keywords

This approach aims to fix the issue at the source, ensuring the parser *always* outputs a `SyntaxKind::Keyword` for control flow words. This would involve a major architectural change.

*   **Sub-option 2a: Lexer Feedback Loop**: The parser would need to send signals back to the lexer to change its state mid-parse (e.g., \"we are now in a procedure, treat `IF` as a keyword\").
    *   **Pros**: Would produce a \"pure\" AST, simplifying the rules.
    *   **Cons**: **Extremely high risk and complexity**. This creates a \"lexer hack\" that is difficult to design, debug, and maintain, and could introduce performance overhead. It would require a significant rewrite of the core parsing engine.

*   **Sub-option 2b: AST Post-Processing**: A second pass would walk the generated AST and mutate it, changing `Word` tokens to `Keyword` tokens based on their context.
    *   **Pros**: Lower risk than a feedback loop as it isolates the logic.
    *   **Cons**: Adds a performance penalty due to the second full pass over the AST and introduces redundant logic that is already present in the word-aware parser rules.

### Final Recommendation

The most effective and lowest-risk solution is **Approach 1: Modify the LT02 rule**.

The parser is correctly handling the complexities of the T-SQL language. The fix should be to make the LT02 rule \"dialect-aware\" enough to handle the AST that the T-SQL parser produces. This approach addresses the problem at the correct layer—the rule interpretation—and avoids a high-risk, complex rewrite of the core parsing engine.
